src/coreclr/jit/importer.cpp

```diff
@@ -4167,15 +4169,87 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
     {
+        unsigned totalSize = info.compCompHnd->getClassSize(fieldClsHnd);
+        unsigned fieldsCnt = info.compCompHnd->getClassNumInstanceFields(fieldClsHnd);
 
-        if (info.compCompHnd->getClassNumInstanceFields(fieldClsHnd) != 1)
+        // For large structs we only want to handle "initialized with zero" case
+        // e.g. Guid.Empty and decimal.Zero static readonly fields.
+        if ((totalSize > TARGET_POINTER_SIZE) || (fieldsCnt != 1))
         {
+            JITDUMP("checking if we can do anything for a large struct ...");
+            const int MaxStructSize = 16;
+            if ((totalSize == 0) || (totalSize > MaxStructSize))
+            {
+                // Limit to 16 bytes for better throughput, larger structs didn't increase the diffs
+                JITDUMP("struct is larger than 16 bytes - bail out.");
+                return nullptr;
+            }
+
+            uint8_t buffer[MaxStructSize] = {0};
+            if (info.compCompHnd->getReadonlyStaticFieldValue(field, buffer, totalSize))
+            {
+                for (unsigned i = 0; i < totalSize; i++)
+                {
+                    if (buffer[i] != 0)
+                    {
+                        // Value is not all zeroes - bail out
+                        JITDUMP("value is not all zeros - bail out.");
+                        return nullptr;
+                    }
+                }
+
+                DWORD typeFlags = info.compCompHnd->getClassAttribs(fieldClsHnd);
+                if (StructHasOverlappingFields(typeFlags) || StructHasCustomLayout(typeFlags))
+                {
+                    // Technically, we only care about holes in a struct, but overlapping fields would complicate
+                    // the holes detection logic.
+                    JITDUMP("struct has complex layout - bail out.");
+                    return nullptr;
+                }
+
+                unsigned totalFldSize = 0;
+                for (unsigned fldIndex = 0; fldIndex < fieldsCnt; fldIndex++)
+                {
+                    CORINFO_FIELD_HANDLE innerField = info.compCompHnd->getFieldInClass(fieldClsHnd, fldIndex);
+                    CORINFO_CLASS_HANDLE innerFieldClsHnd;
+                    var_types            fieldVarType =
+                        JITtype2varType(info.compCompHnd->getFieldType(innerField, &innerFieldClsHnd, fieldClsHnd));
+
+                    if (!varTypeIsIntegral(fieldVarType))
+                    {
+                        // A conservative check since a proper one is more complicated and doesn't bring new diffs
+                        JITDUMP("struct has non-primitive fields - bail out.");
+                        return nullptr;
+                    }
+                    totalFldSize += genTypeSize(fieldVarType);
+                }
+
+                if (totalFldSize != totalSize)
+                {
+                    JITDUMP("struct has holes - bail out.");
```

Comment:
Is this for correctness or CQ?

Reply:
Correctness, I assume I can remove all of that legally, right? since even in case of holes, the data in the holes is 0 too

Reply:
> I assume I can remove all of that legally, right? since even in case of holes, the data in the holes is 0 too

Yes, InitBlk forms initialize the whole struct.